dali/kernels/imgproc/warp_cpu.h

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_WARP_CPU_H_
+#define DALI_KERNELS_IMGPROC_WARP_CPU_H_
+
+#include "dali/core/common.h"
+#include "dali/core/geom/vec.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/imgproc/warp/mapping_traits.h"
+#include "dali/kernels/imgproc/sampler.h"
+#include "dali/kernels/imgproc/warp/map_coords.h"
+
+namespace dali {
+namespace kernels {
+
+/// @remarks Assume HWC layout
+template <typename _Mapping, int ndim, typename _OutputType, typename _InputType,
+          typename _BorderType>
+class WarpCPU {
+ public:
+  static constexpr int spatial_ndim = ndim;
+  static constexpr int tensor_ndim = ndim + 1;
+  static constexpr int channel_dim = ndim;
+
+  static_assert(spatial_ndim == 2, "Not implemented for spatial_ndim != 2");
+
+  using Mapping = _Mapping;
+  using OutputType = _OutputType;
+  using InputType = _InputType;
+  using BorderType = _BorderType;
+  using MappingParams = warp::mapping_params_t<Mapping>;
+
+  KernelRequirements Setup(
+      KernelContext &context,
+      const InTensorCPU<InputType, tensor_ndim> &input,
+      const MappingParams &mapping_params,
+      const TensorShape<spatial_ndim> &out_size,
+      DALIInterpType interp = DALI_INTERP_LINEAR,
+      const BorderType &border = {}) {
+    KernelRequirements req;
+    req.output_shapes.resize(1);
+    auto out_shape = shape_cat(out_size, input.shape[channel_dim]);
+    req.output_shapes[0] = TensorListShape<tensor_ndim>({out_shape});
```

Comment:
```suggestion
    req.output_shapes.push_back(TensorListShape<tensor_ndim>({out_shape}));
```
and remove the resize(1) call?